components/feature/pwa/src/main/java/mozilla/components/feature/pwa/intent/WebAppIntentProcessor.kt

```diff
@@ -0,0 +1,53 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.pwa.intent
+
+import android.content.Intent
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.Session.Source
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.browser.session.intent.IntentProcessor
+import mozilla.components.browser.session.intent.putSessionId
+import mozilla.components.concept.engine.EngineSession
+import mozilla.components.feature.pwa.ManifestStorage
+import mozilla.components.feature.pwa.ext.putWebAppManifest
+import mozilla.components.feature.pwa.ext.toCustomTabConfig
+import mozilla.components.feature.session.SessionUseCases
+import mozilla.components.support.utils.toSafeIntent
+
+class WebAppIntentProcessor(
+    private val sessionManager: SessionManager,
+    private val loadUrlUseCase: SessionUseCases.DefaultLoadUrlUseCase,
+    private val storage: ManifestStorage
+) : IntentProcessor {
+
+    override fun matches(intent: Intent) =
+        intent.toSafeIntent().action == ACTION_VIEW_PWA
+
+    override suspend fun process(intent: Intent): Boolean {
+        val url = intent.toSafeIntent().dataString
+
+        return if (!url.isNullOrEmpty() && matches(intent)) {
+            val webAppManifest = storage.loadManifest(url) ?: return false
+
+            val session = Session(url, private = false, source = Source.HOME_SCREEN)
+            session.webAppManifest = webAppManifest
+            session.customTabConfig = webAppManifest.toCustomTabConfig()
+
+            sessionManager.add(session)
```

Comment:
By adding the custom tab config here we are piggy backing on the fact that custom tabs are excluded from most of the browser UI. I think we should keep them as completely separate entities that may be handled differently and that may have different lifetimes. But I also agree that we shouldn't start dealing with that in `browser-session` - but definitely in `browser-state`.